drivers/dht/dht.c

```diff
@@ -19,201 +20,260 @@
  * @author      Ludwig Knüpfer <ludwig.knuepfer@fu-berlin.de>
  * @author      Christian Mehlis <mehlis@inf.fu-berlin.de>
  * @author      Hauke Petersen <hauke.petersen@fu-berlin.de>
+ * @author      Hugues Larrive <hugues.larrive@pm.me>
  *
  * @}
  */
 
+#include <errno.h>
 #include <stdint.h>
 #include <string.h>
 
-#include "log.h"
 #include "assert.h"
-#include "xtimer.h"
-#include "timex.h"
-#include "periph/gpio.h"
-
 #include "dht.h"
 #include "dht_params.h"
+#include "log.h"
+#include "periph/gpio.h"
+#include "time_units.h"
+#include "ztimer.h"
 
 #define ENABLE_DEBUG            0
 #include "debug.h"
 
 /* Every pulse send by the DHT longer than 40µs is interpreted as 1 */
-#define PULSE_WIDTH_THRESHOLD   (40U)
-/* If an expected pulse is not detected within 1000µs, something is wrong */
-#define TIMEOUT                 (1000U)
-/* The DHT sensor cannot measure more than once a second */
-#define DATA_HOLD_TIME          (US_PER_SEC)
-/* The start signal by pulling data low for at least 18ms and then up for
- * 20-40µs*/
-#define START_LOW_TIME          (20U * US_PER_MS)
-#define START_HIGH_TIME         (40U)
-
-static inline void _reset(dht_t *dev)
+#define READ_THRESHOLD          (40U)
+/* If an expected pulse is not detected within 85µs, something is wrong */
+#define SPIN_TIMEOUT            (85U)
+/* The start signal by pulling data low for at least 18 ms for DHT11, at
+ * most 20 ms (AM2301 / DHT22 / DHT21). Then release the bus and the
+ * sensor should respond by pulling data low for 80 µs, then release for
+ * 80µs before start sending data. */
+#define START_LOW_TIME          (19U * US_PER_MS)
+#define START_THRESHOLD         (75U)
+/* DHTs have to wait for power 1 or 2 seconds depending on the model */
+#define POWER_WAIT_TIMEOUT      (2U * US_PER_SEC)
+
+enum {
+    BYTEPOS_HUMIDITY_HIGH = 0,
+    BYTEPOS_HUMIDITY_LOW = 1,
+    BYTEPOS_TEMPERATURE_HIGH = 2,
+    BYTEPOS_TEMPERATURE_LOW = 3,
+    BYTEPOS_CHECKSUM = 4,
+};
+
+struct dht_data {
+    gpio_t pin;
+    gpio_mode_t in_mode;
+    uint8_t data[5];
+    int8_t bit_pos;
+    uint8_t bit;
+};
+
+static void _wait_for_level(gpio_t pin, bool expected, uint32_t start)
+{
+    /* Calls to ztimer_now() can be relatively slow on low end platforms.
+     * Mixing in a busy down-counting loop solves issues e.g. on AVR boards. */
+    uint8_t pre_timeout = 0;
+    while (((bool)gpio_read(pin) != expected) && (++pre_timeout
+        || ztimer_now(ZTIMER_USEC) < start + SPIN_TIMEOUT)) {}
```

Comment:
```suggestion
    while (((bool)gpio_read(pin) != expected) &&
            (++pre_timeout || ztimer_now(ZTIMER_USEC) < start + SPIN_TIMEOUT)) { }
```

Personally, I think breaking in the middle of the second operand to the `&&` operator is hard to read.

If you agree, please squash/amend directly.

Reply:
No, I prefer to keep it as is because the conditions in the first line are processed 256 times before the second line is processed for the first time, which will only happen on the fastest CPUs.